bitcoin/test/fuzz/miniscript_random.cpp

```diff
@@ -241,93 +247,423 @@ std::optional<uint32_t> ConsumeTimeLock(FuzzedDataProvider& provider) {
+            }
+            if (useful_types.size() == set_size) break;
+        }
+        // Remove all rules that construct uninteresting types.
+        for (auto type_it = table.begin(); type_it != table.end();) {
+            if (useful_types.count(type_it->first) == 0) {
+                type_it = table.erase(type_it);
+            } else {
+                ++type_it;
+            }
+        }
+
+        /* Find which types are constructible. A type is constructible if there is a leaf
+         * node recipe for constructing it, or a recipe whose subnodes are all constructible.
+         * Types can be non-constructible because they have no recipes to begin with,
+         * because they can only be constructed using recipes that involve otherwise
+         * non-constructible types, or because they require infinite recursion. */
+        std::set<Type> constructible_types{};
+        auto known_constructible = [&](Type type) { return constructible_types.count(type) != 0; };
+        // Find the transitive closure by adding types until the set of types does not change.
+        while (true) {
+            size_t set_size = constructible_types.size();
+            // Iterate over all types we have recipes for.
+            for (const auto& [type, recipes] : table) {
+                if (!known_constructible(type)) {
+                    // For not (yet known to be) constructible types, iterate over their recipes.
+                    for (const auto& [_, subt] : recipes) {
+                        // If any recipe involves only (already known to be) constructible types,
+                        // add the recipe's type to the set.
+                        if (std::all_of(subt.begin(), subt.end(), known_constructible)) {
+                            constructible_types.insert(type);
+                            break;
+                        }
+                    }
+                }
+            }
+            if (constructible_types.size() == set_size) break;
+        }
+        for (auto type_it = table.begin(); type_it != table.end();) {
+            // Remove all recipes which involve non-constructible types.
+            type_it->second.erase(std::remove_if(type_it->second.begin(), type_it->second.end(),
+                [&](const recipe& rec) {
+                    return !std::all_of(rec.second.begin(), rec.second.end(), known_constructible); 
+                }), type_it->second.end());
+            // Delete types entirely which have no recipes left.
+            if (type_it->second.empty()) {
+                type_it = table.erase(type_it);
+            } else {
+                ++type_it;
+            }
+        }
+
+        for (auto& [type, recipes] : table) {
+            // Sort recipes for determinism, and place those using fewer subnodes first.
+            // This avoids runaway expansion (when reaching the end of the fuzz input,
+            // all zeroes are read, resulting in the first available recipe being picked).
```

Comment:
This means we need to be wary not to introduce recipes with "inter dependencies" as the ones with fewer subnodes, right?

For instance if we were to only read zeroes, and the smallest recipe to create a `Bd` was `and_b(Bd,Wd)` then we'd have an infinite loop. As far as i can tell they would not be removed above as `Bd` would be `known_constructible` by other means?

Reply:
Indeed - if the smallest one was recursive like that, we'd have a problem. That happens to not be the case for any type, currently.

I think if the types/fragments ever change, and this were to become the case, that can still be addressed then. It should be very obvious once it happens, I think.